From a639cc5b184f32f0a69ee8621075849b7bfddc3a Mon Sep 17 00:00:00 2001
From: De Huo <De.Huo@windriver.com>
Date: Thu, 8 Aug 2019 18:59:06 +0800
Subject: [PATCH] Rule cklib_accounts_password_pam_minclass add oval check

For rule cracklib_accounts_password_pam_minclass add oval check

Upstream-Status: Pending

Signed-off-by: De Huo <De.Huo@windriver.com>
---
 .../bash/wrlinux1019.sh                            | 14 ++++++++++
 .../oval/wrlinux1019.xml                           | 32 ++++++++++++++++++++++
 2 files changed, 46 insertions(+)
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pamcracklib/cracklib_accounts_password_pam_minclass/bash/wrlinux1019.sh
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pamcracklib/cracklib_accounts_password_pam_minclass/oval/wrlinux1019.xml

diff --git a/linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pamcracklib/cracklib_accounts_password_pam_minclass/bash/wrlinux1019.sh b/linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pamcracklib/cracklib_accounts_password_pam_minclass/bash/wrlinux1019.sh
new file mode 100644
index 0000000..42ec140
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pamcracklib/cracklib_accounts_password_pam_minclass/bash/wrlinux1019.sh
@@ -0,0 +1,14 @@
+# platform = multi_platform_wrlinux
+
+. /usr/share/scap-security-guide/remediation_functions
+populate var_password_pam_minclass
+
+if grep -q '^password.*pam_cracklib.so' /etc/pam.d/common-password; then
+    if grep -q '^password.*pam_cracklib.so.*minclass=' /etc/pam.d/common-password; then
+        sed -i --follow-symlinks "/^password.*pam_cracklib.so/ s/\(minclass\)=\D+/\1=$var_password_pam_minclass/" /etc/pam.d/common-password
+    else
+        sed -i --follow-symlinks "/^password.*pam_cracklib.so/ s/$/ minclass=$var_password_pam_minclass/" /etc/pam.d/common-password
+    fi
+else
+    echo -e "password\trequired\tpam_cracklib.so minclass=$var_password_pam_minclass/" >>/etc/pam.d/common-password
+fi
diff --git a/linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pamcracklib/cracklib_accounts_password_pam_minclass/oval/wrlinux1019.xml b/linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pamcracklib/cracklib_accounts_password_pam_minclass/oval/wrlinux1019.xml
new file mode 100644
index 0000000..2d6b995
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pamcracklib/cracklib_accounts_password_pam_minclass/oval/wrlinux1019.xml
@@ -0,0 +1,32 @@
+<def-group>
+ <definition class="compliance" id="cracklib_accounts_password_pam_minclass" version="1">
+    <metadata>
+      <title>Set Password minclass Requirements</title>
+      <affected family="unix">
+        <platform>multi_platform_wrlinux</platform>
+      </affected>
+      <description>The password minclass should meet minimum requirements using pam_cracklib</description>
+    </metadata>
+    <criteria>
+      <criterion comment="Conditions for minclass are satisfied" test_ref="test_accounts_password_pam_cracklib_minclass" />
+    </criteria>
+  </definition>
+
+  <ind:textfilecontent54_test check="all" comment="check the configuration of /etc/pam.d/common-password" id="test_accounts_password_pam_cracklib_minclass" version="1">
+    <ind:object object_ref="object_accounts_password_pam_cracklib_minclass" />
+    <ind:state state_ref="state_accounts_password_pam_cracklib_minclass" />
+  </ind:textfilecontent54_test>
+  <ind:textfilecontent54_object id="object_accounts_password_pam_cracklib_minclass" version="1">
+    <ind:path>/etc/pam.d</ind:path>
+    <ind:filename>common-password</ind:filename>
+    <ind:pattern operation="pattern match">^password[\s]+(?:requisite|required)[\s]+(?:/lib/security/(?:\$ISA/)*)*pam_cracklib.so[\w_\.\-=\s]*[\s]+minclass=(-?\d+)(?:[\s]|$)</ind:pattern>
+    <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>
+  </ind:textfilecontent54_object>
+  <ind:textfilecontent54_state id="state_accounts_password_pam_cracklib_minclass" version="1">
+    <ind:instance datatype="int">1</ind:instance>
+    <ind:subexpression datatype="int" operation="greater than or equal" var_ref="var_password_pam_minclass" />
+  </ind:textfilecontent54_state>
+
+  <external_variable comment="External variable for pam_cracklib minclass" datatype="int" id="var_password_pam_minclass" version="1" />
+
+</def-group>
-- 
1.9.1

